buildSrc/src/main/kotlin/Java9Modularity.kt

```diff
@@ -0,0 +1,65 @@
+import org.gradle.api.JavaVersion
+import org.gradle.api.Project
+import org.gradle.api.tasks.bundling.Jar
+import org.gradle.kotlin.dsl.findByType
+import org.jetbrains.kotlin.gradle.dsl.KotlinMultiplatformExtension
+import org.jetbrains.kotlin.gradle.targets.jvm.KotlinJvmTarget
+import java.util.spi.ToolProvider
+
+object Java9Modularity {
+    @JvmStatic
+    fun Project.configureJava9ModuleInfo(moduleName: String) {
+        val multiplatform = extensions.findByType<KotlinMultiplatformExtension>() ?: return
+        val jvmTargets = multiplatform.targets.filterIsInstance<KotlinJvmTarget>().filter { it.withJavaEnabled }
+        if (jvmTargets.isEmpty()) {
+            logger.warn("No Kotlin JVM targets found with Java enabled, can't configure Java 9 source set!")
+        }
+        jvmTargets.forEach { target ->
+            val mainCompilation = target.compilations.getByName("main")
+            val inputClasspath = mainCompilation.compileKotlinTask.classpath
+            val outputClassesDir = mainCompilation.compileKotlinTask.destinationDir
+            val outputJarTask = tasks.getByName(target.artifactsTaskName) as Jar
+
+            val java9SourceSet = multiplatform.sourceSets.create("${target.targetName}Java9")
+            val java9Compilation = target.compilations.create("java9") {
```

Comment:
Am I correct that this compilation is always created and compiled on every build?

Reply:
Yes, if the method `configureJava9ModuleInfo` is called on a project It will always setup the sourceset and it's associated compilation.